### PR TITLE
feat(mock): allow matching the complete header list

### LINKED ISSUE #1269: Mocks should allow to match headers that are meant to be missing
## This would solve...

Being able to match only when a given header is not present.

## The implementation should look like...

I dunno, ...
- prefix the header name with a unique symbol outside the range of allowed header names to indicate it's meant to be missing in a to be matched request?
- special symbol instead of a string|regex|function as a matcher to be provided
- function matcher would get undefined passed as a value?

## I have also considered...

I've tried to match the missing header with a function that checks that the value is undefined but the matcher doesn't even run because the header is not in the request.

### PR DESCRIPTION
**Pull request by panva**
resolves #1269 and (subjectively) a preferred alternative to #1274

### CODE REVIEW

**Comment by panva** [comment]
Might need to use a regular object instead of Headers for node < 16 compat. Happy to do so given otherwise non blocking feedback.

**Comment by mcollina** [comment]
code looks great!

**Review by mcollina — APPROVED** [review]
lgtm

File: types/mock-interceptor.d.ts [hunk 1]
```diff
@@ -49,7 +49,7 @@ declare namespace MockInterceptor {
     /** Body to intercept on. */
     body?: string | RegExp | ((body: string) => boolean);
     /** Headers to intercept on. */
-    headers?: Record<string, string | RegExp | ((body: string) => boolean)>;
+    headers?: Record<string, string | RegExp | ((body: string) => boolean)> | ((headers: Record<string, string>) => boolean);
```
Comment (mcollina) [inline types/mock-interceptor.d.ts:52]: Could you add an assertion for this type? We use tsd?

Comment (panva) [reply types/mock-interceptor.d.ts:52]: Sure, where?

Comment (mcollina) [reply types/mock-interceptor.d.ts:52]: test/types/mock-interceptor.test-d.ts

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**